rest/route/host_spawn.go

```diff
@@ -1010,6 +1011,93 @@ func (h *hostExtendExpirationHandler) Run(ctx context.Context) gimlet.Responder
 	return gimlet.NewJSONResponse(struct{}{})
 }
 
+////////////////////////////////////////////////////////////////////////
+//
+// POST /rest/v2/hosts/{host_id}/run_command
+//
+type hostRunCommand struct {
+	sc  data.Connector
+	env evergreen.Environment
+
+	hostID  string
+	command string
+}
+
+func makeHostRunCommand(sc data.Connector, env evergreen.Environment) gimlet.RouteHandler {
+	return &hostRunCommand{
+		sc:  sc,
+		env: env,
+	}
+}
+
+func (h *hostRunCommand) Factory() gimlet.RouteHandler {
+	return &hostRunCommand{
+		sc:  h.sc,
+		env: h.env,
+	}
+}
+
+func (h *hostRunCommand) Parse(ctx context.Context, r *http.Request) error {
+	var err error
+	h.hostID, err = validateID(gimlet.GetVars(r)["host_id"])
+	if err != nil {
+		return errors.Wrap(err, "invalid host id")
+	}
+
+	hostCommand := model.APIHostCommand{}
+	if err = util.ReadJSONInto(util.NewRequestReader(r), &hostCommand); err != nil {
+		return errors.Wrap(err, "can't read host command from json")
+	}
+	h.command = hostCommand.Command
+
+	return nil
+}
+
+func (h *hostRunCommand) Run(ctx context.Context) gimlet.Responder {
+	user := MustHaveUser(ctx)
+	host, err := host.FindOneId(h.hostID)
+	if err != nil {
+		return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "can't get host"))
+	}
+	if host == nil {
+		response := gimlet.MakeJSONErrorResponder(errors.Errorf("no host found for id '%s'", h.hostID))
+		_ = response.SetStatus(http.StatusNotFound)
+		return response
+	}
+	if host.StartedBy != user.Id {
+		response := gimlet.MakeJSONErrorResponder(errors.New("Unauthorized"))
+		_ = response.SetStatus(http.StatusUnauthorized)
+		return response
+	}
+	if host.Status != evergreen.HostRunning {
+		return gimlet.MakeJSONInternalErrorResponder(errors.Errorf("can't run command on host with status '%s'", host.Status))
+	}
+
+	if host.Distro.BootstrapSettings.Method == distro.BootstrapMethodUserData {
```

Comment:
Whether or not to or run something via a Jasper service or regular SSH should depend on the `(host.Host).Distro.BootstrapSettings.Communication` setting, which can be legacy (i.e. we only support regular SSH) or new (i.e. we can connect to the host's Jasper service). This check should instead be `if host.JasperCommunication()`, because inside `RunJasperProcess`, it calls `(*host.Host).JasperClient()`, which connects to any host assuming it is not using legacy provisioning (i.e. has Jasper running).

Reply:
done.